frame/vesting/src/lib.rs

```diff
@@ -61,41 +60,139 @@ use frame_support::traits::{
+			locked: Balance,
+			per_block: Balance,
+			starting_block: BlockNumber,
+		) -> Result<VestingInfo<Balance, BlockNumber>, Error<T>> {
+			Self::validate_params(locked, per_block, starting_block)?;
+			let per_block = if per_block > locked { locked } else { per_block };
+			Ok(VestingInfo { locked, per_block, starting_block })
+		}
+
+		/// Validate parameters for `VestingInfo`.
+		pub fn validate_params<T: Config>(
+			locked: Balance,
+			per_block: Balance,
+			_starting_block: BlockNumber,
+		) -> Result<(), Error<T>> {
+			ensure!(!locked.is_zero() && !per_block.is_zero(), Error::<T>::InvalidScheduleParams);
+			let min_transfer: u32 = T::MinVestedTransfer::get().try_into().unwrap_or(u32::MAX);
+			let min_transfer = Balance::from(min_transfer);
+			ensure!(locked >= min_transfer, Error::<T>::AmountLow);
+			Ok(())
+		}
+
+		/// Instantiate a new `VestingInfo` without param validation. Useful for
+		/// mocking bad inputs in testing.
+		pub fn unsafe_new(
+			locked: Balance,
+			per_block: Balance,
+			starting_block: BlockNumber,
+		) -> VestingInfo<Balance, BlockNumber> {
+			VestingInfo { locked, per_block, starting_block }
+		}
+
+		/// Locked amount at genesis.
+		pub fn locked(&self) -> Balance {
+			self.locked
+		}
+
+		/// Amount that gets unlocked every block after `starting_block`.
+		pub fn per_block(&self) -> Balance {
+			self.per_block
+		}
+
+		/// Starting block for unlocking(vesting).
+		pub fn starting_block(&self) -> BlockNumber {
+			self.starting_block
+		}
+
+		/// Amount locked at block `n`.
+		pub fn locked_at<BlockNumberToBalance: Convert<BlockNumber, Balance>>(
+			&self,
+			n: BlockNumber,
+		) -> Balance {
+			// Number of blocks that count toward vesting
+			// Saturating to 0 when n < starting_block
+			let vested_block_count = n.saturating_sub(self.starting_block);
+			let vested_block_count = BlockNumberToBalance::convert(vested_block_count);
+			// Return amount that is still locked in vesting
+			let maybe_balance = vested_block_count.checked_mul(&self.per_block);
+			if let Some(balance) = maybe_balance {
+				self.locked.saturating_sub(balance)
+			} else {
+				Zero::zero()
+			}
+		}
+
+		/// Block number at which the schedule ends
+		pub fn ending_block<BlockNumberToBalance: Convert<BlockNumber, Balance>>(&self) -> Balance {
+			let starting_block = BlockNumberToBalance::convert(self.starting_block);
+			let duration = if self.per_block > self.locked {
+				// If `per_block` is bigger than `locked`, the schedule will end
+				// the block after starting
+				1u32.into()
+			} else if self.per_block.is_zero() {
+				// Check for div by 0 errors, which should only be from legacy
+				// vesting schedules since new ones are validated for this.
+				self.locked
+			} else {
+				let has_remainder = !(self.locked % self.per_block).is_zero();
+				let maybe_duration = self.locked / self.per_block;
+				if has_remainder {
+					maybe_duration + 1u32.into()
+				} else {
+					maybe_duration
+				}
```

Comment:
```suggestion
				self.locked / self.per_block + if self.locked % self.per_block != Zero::zero() { 
					1u32.into() 
				} else { 
					Zero::zero() 
				}
```

simplify

Reply:
in the normal case we expect `maybe_duration` to be greater than zero, so we need to take it into account

Reply:
I did; it's before the `if` clause there. Just inlined.

Reply:
089f5da417f42c0bbddb30147b4be818eddb2911